components/region_cache_memory_engine/src/background.rs

```diff
@@ -1027,6 +1071,84 @@ impl RunnableWithTimer for BackgroundRunner {
     }
 }
 
+pub struct DeleteRangeRunner {
+    engine: Arc<RwLock<RangeCacheMemoryEngineCore>>,
+    // It is possible that when `DeleteRangeRunner` begins to delete a range, the range is being
+    // written by apply threads. In that case, we have to delay the delete range task to avoid race
+    // condition between them. Periodically, these delayed ranges will be checked to see if it is
+    // ready to be deleted.
+    delay_ranges: Vec<CacheRange>,
+}
+
+impl DeleteRangeRunner {
+    fn new(engine: Arc<RwLock<RangeCacheMemoryEngineCore>>) -> Self {
+        Self {
+            engine,
+            delay_ranges: vec![],
+        }
+    }
+
+    fn delete_ranges(&mut self, ranges: &[CacheRange]) {
+        let skiplist_engine = self.engine.read().engine();
+        for r in ranges {
+            println!("delete range {:?}", r);
+            skiplist_engine.delete_range(r);
+        }
+        self.engine
+            .write()
+            .mut_range_manager()
+            .on_delete_ranges(ranges);
+        #[cfg(test)]
+        flush_epoch();
```

Comment:
I'm curious why do we have to flush_epoch here?

Reply:
In crossbeam skiplist, after delete, the memory free is not performed immedietly, it is buffered and delayed until it is guaranteed that no other threads will read the memory. The flush here is to accelerate this process.